admin/cincoctrl/cincoctrl/findingaids/models.py

```diff
@@ -24,61 +32,74 @@
-    date_created = DateTimeField(auto_now_add=True) # auto
-    date_updated = DateTimeField(auto_now=True) # auto
+    ) # system-set
+    collection_title = CharField(max_length=255) # extracted from ead if present
+    collection_number = CharField(max_length=255) # extracted from ead if present
+    ead_file = FileField(upload_to="ead/",
+                         validators=[FileExtensionValidator(['xml']), validate_ead],
+                         null=True, blank=True)
+    record_type = CharField(max_length=10, choices=RECORD_TYPES) # system-set
+    status = CharField(max_length=50, default="imported", choices=STATUSES) # system-set
+    date_created = DateTimeField(auto_now_add=True) # auto-assigned
+    date_updated = DateTimeField(auto_now=True) # auto-updated
+
+    def extract_ead_fields(self):
+        with self.ead_file.open('rb') as f:
+            root = ET.parse(f).getroot()
+        title_node = root.find("./archdesc/did/unittitle")
+        number_node = root.find("./archdesc/did/unitid")
+        return title_node.text, number_node.text
+
+    def save(self, *args, **kwargs):
+        if not self.ark:
+            self.ark = uuid.uuid4()
+        super().save(*args, **kwargs)
+        if self.record_type != 'express':
+            self.collection_title, self.collection_number = self.extract_ead_fields()
+            super().save(*args, **kwargs)
+
+    def get_absolute_url(self) -> str:
+        if self.record_type == 'express':
+            return reverse("findingaids:view_record_express",
+                           kwargs={"pk": self.expressrecord.pk,
+                                   "repo_slug": self.repository.code})
+        else:
+            return reverse("findingaids:view_record",
+                           kwargs={"pk": self.id,
+                                   "repo_slug": self.repository.code})
 
     def __str__(self):
         return self.collection_title
 
 class SupplementaryFile(models.Model):
     finding_aid = ForeignKey("FindingAid", on_delete=models.CASCADE)
-    file = ForeignKey('File', on_delete=models.CASCADE)
+    pdf_file = FileField(upload_to="pdf/", validators=[FileExtensionValidator(['pdf'])])
     date_created = DateTimeField(auto_now_add=True)
     date_updated = DateTimeField(auto_now=True)
 
     def __str__(self):
-        return f"{self.finding_aid} / {self.file}"
-
-class File(models.Model):
-    file = FileField()
+        return f"{self.finding_aid} / {self.pdf_file}"
 
 class ExpressRecord(models.Model):
-    finding_aid = ForeignKey("FindingAid", on_delete=models.CASCADE)
+    finding_aid = OneToOneField("FindingAid", on_delete=models.CASCADE)
     title_filing = CharField("Collection Title (Filing)", max_length=255)
-    local_identifier = CharField("Collection Identifier/Call Number", max_length=255)
     date_dacs = CharField("Collection Date", max_length=128)
+    # TODO: add validator for ISO date range
```

Comment:
DACS [Describing Archives: A content standard](https://github.com/saa-ts-dacs/dacs)

This was what the field was called in the old Record Express.  We have the two date fields, ISO should actually be a validatable ISO date (but it allows date ranges that may only be years so we'll have to write a custom validator).  The date_dacs follows general guidelines but is much looser and should be human readable like "circa 2003-2020" or something.

Reply:
I don't see the link in config/__init__.py that you are referring to. I wouldn't call the config directory an app really, a module,  I guess.  It doesn't define any models or views.  I like this layout because it doesn't require one of the apps to be the "controller" so when we add new apps we add them only in config directory (installed_apps and urls) not in another app.

As for repository, I guess it's a dependency thing.  I want UserRole and permissions stuff in the users directory but permissions depend on being associated with a repository.

Reply:
Ahh, thanks for the link to DACS! I am sensitive to singular/plural distinctions in code and was unfamiliar with the acronym. I think I have a slight preference for a standards prefixes rather than suffix, as in `iso_date` and `dacs_date`. 

This is the link that doesn't resolve: https://github.com/ucldc/cinco/blob/AT-cinco-10-admin-forms/admin/cincoctrl/cincoctrl/contrib/__init__.py - `contrib` does have migrations, but it looks like a sites module or something? I'm not sure what you mean by requiring one of the apps to the be "controller"?

Reply:
Ohhh, sorry I misunderstood your message because there's both a `config` and a `contrib` directory. I'm talking about the `contrib` directory, not the `config` directory. 

I'm still not totally sure what the `contrib` directory is doing, but yeah, I see what you mean re: this layout. The classic Django "project" directory sits adjacent to the app directories, but the "project" directory typically contains `settings.py`, the root `urls.py`, `wsgi.py`, etc (I'm guessing this is what you mean by "controller"?), while the app directories contain views and models. When looking at a new Django project, it's a bit confusing to figure out which directory is the one that contains these project-level settings as opposed to all the other apps. This layout puts project-level configuration in the `config` directory, and nests apps together in the `cincoctrl` directory?

Reply:
I updated the link in the readme PR